gix-worktree/src/diff.rs

```diff
@@ -0,0 +1,292 @@
+            || stat.mtime.nsecs != mtime.subsec_nanos()
+            || stat.ctime.secs
+                != ctime
+                    .as_secs()
+                    .try_into()
+                    .expect("by 2038 we found a solution for this")
+            || stat.ctime.nsecs != ctime.subsec_nanos();
+
+        Ok(Self {
+            mode_change,
+            stat_changed,
+            data_changed,
+        })
+    }
+
+    /// Marks this entries stats as changed if there is a potential fs race condition
+    pub fn detect_racy_stat(&mut self, index: &index::State, index_entry: &index::Entry) {
+        self.stat_changed = self.stat_changed || index_entry.stat.mtime >= index.timestamp()
+    }
+
+    /// returns true if this entry has any changes
+    /// usually `detect_racy_stat` should be called first to avoid race condition
+    pub fn changed(&self) -> bool {
+        self.mode_change.is_some() || self.stat_changed || self.data_changed
+    }
+
+    /// Reads the worktree file from the disk and compares it to
+    /// the index entries oid to check if the actual data of the file is changed
+    /// and sets [`Entry::data_changed`] accordingly
+    pub fn compare_data(
+        &mut self,
+        worktree_path: &Path,
+        index_entry: &index::Entry,
+        buf: &mut Vec<u8>,
+        capabilities: &fs::Capabilities,
+    ) -> Result<(), read::Error> {
+        if self.mode_change.is_some() || !self.stat_changed || self.data_changed {
+            return Ok(());
+        }
+        let data = read_blob_to_buf_with_meta(
+            worktree_path,
+            index_entry.mode.contains(index::entry::Mode::SYMLINK),
+            buf,
+            capabilities,
+        )?;
+        let header = loose_header(gix_object::Kind::Blob, data.len());
+        let hash_changed = match index_entry.id {
+            ObjectId::Sha1(entry_hash) => {
+                let mut file_hash = hash::Sha1::default();
+                file_hash.update(&header);
+                file_hash.update(&data);
+                let file_hash = file_hash.digest();
+                entry_hash != file_hash
+            }
+        };
+        self.data_changed = hash_changed;
+        Ok(())
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Could not convert path to UTF8 {path}")]
+    IllformedUtf8 { path: BString },
+    #[error("The clock was off when reading file related metadata after updating a file on disk")]
+    Time(#[from] std::time::SystemTimeError),
+    #[error("IO error while writing blob or reading file metadata or changing filetype")]
+    Io(#[from] io::Error),
+}
+
+#[derive(Clone, Debug)]
+/// A change between the index and the worktree computed by [`compate_to_index`]
+pub struct Change<'a> {
+    /// The index entry that changed
+    pub index_entry: &'a index::Entry,
+    /// The on-disk worktree path corresponding to this entry
+    pub worktree_path: PathBuf,
```

Comment:
this is an absolute path tough. so basically `worktree.join(relative_path)` the join always allocated anwyay so I thaught this was the best I could come up with. That said we likely can do better here once I used the FS cache. For now I wouldn't want to implement an in place join since it will be replaced by the FS cache anyways

Reply:
> That said we likely can do better here once I used the FS cache. 

Yes, let's keep that in mind. With the `fs::Cache` you naturally get the absolute path and it reuses the allocation. Then we just have to stop storing it separately.